scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +160,341 @@
+            (throw (ex-info "dynamic-cluster-config? is set to false in cluster definition template but we are trying to add a cluster dynamically"
+                            {:template (:template config)})))
+        full-cluster-config (-> (:config cluster-definition-template) (merge config) (assoc :dynamic-cluster-config? true))
+        cluster (resolved full-cluster-config {:exit-code-syncer-state @exit-code-syncer-state-atom})]
+    (initialize-cluster cluster {:pool-name->fenzo @pool-name->fenzo-atom})))
+
+(defn execute-update!
+  "Attempt to execute a valid cluster configuration update.
+  Change cluster state if there is an existing cluster in memory.
+  Create clusters if there is an active update an no corresponding cluster in memory. This can happen at startup or
+  when a brand new cluster is added.
+  Reflect updates in the database."
+  [conn
+   {:keys [valid? differs? active?] {:keys [name state state-locked?] :as goal-config} :goal-config}
+   current-in-mem-configs]
+  {:pre [valid?]}
+  (try
+    (let [{:keys [state-atom state-locked?-atom] :as cluster} (@cluster-name->compute-cluster-atom name)
+          current-in-mem-config (current-in-mem-configs name)]
+      ; TODO: keep these checks? check some other way?
+      (when (and current-in-mem-config (not (and cluster state-atom state-locked?-atom)))
+        (throw (ex-info "We know an in-memory config but we don't know the corresponding in-memory cluster.
+        This should never happen, since cluster-name->compute-cluster-atom should be locked"
+                        {:current-in-mem-config current-in-mem-config
+                         :cluster cluster})))
+      (when (and (or cluster state-atom state-locked?-atom) (not current-in-mem-config))
+        (throw (ex-info "We know an in-memory cluster but we don't know the corresponding in-memory config.
+        This should never happen, since cluster-name->compute-cluster-atom should be locked"
+                        {:current-in-mem-config current-in-mem-config
+                         :cluster cluster})))
+      (if current-in-mem-config
+        (do
+          (reset! state-atom state)
+          (reset! state-locked?-atom state-locked?))
+        (when active? (initialize-cluster! goal-config))))
+    (when differs?
+      ; :db.unique/identity gives upsert behavior (update on insert), so we don't need to specify an entity ID when updating
+      @(d/transact conn [(assoc (compute-cluster-config->compute-cluster-config-ent goal-config)
+                           :db/id (d/tempid :db.part/user))]))
+    {:update-succeeded true}
+    (catch Throwable t
+      (log/error t "Failed to update cluster" goal-config)
+      {:update-succeeded false :error-message (.toString t)})))
+
+(defn update-compute-clusters-helper
+  "Helper function for update-compute-cluster and update-compute-clusters. See those functions for full description."
```

Comment:
I'm 100% OK with the lock being grabbed in two locations (instead of one). This smells like those 4 functions (which can now be independently tested!), and then we have no closures being passed around.

Reply:
I like @scrosby's proposal.

Reply:
updated